components/object_store/src/obkv/mod.rs

```diff
@@ -0,0 +1,940 @@
+            }));
+        }
+
+        let iter = futures::stream::iter(meta_list.into_iter());
+        debug!(
+            "ObkvObjectStore list operation, prefix:{path}, cost:{}",
+            instant.elapsed().as_millis()
+        );
+        Ok(iter.boxed())
+    }
+
+    /// List all the objects and common paths(directories) with the given
+    /// prefix. Prefixes are evaluated on a path segment basis, i.e.
+    /// `foo/bar/` is a prefix of `foo/bar/x` but not of `foo/bar_baz/x`.
+    /// see detail in: https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html
+    async fn list_with_delimiter(&self, prefix: Option<&Path>) -> Result<ListResult> {
+        let instant = Instant::now();
+
+        let path = Self::normalize_path(prefix);
+        let metas = self
+            .meta_manager
+            .list(&path)
+            .await
+            .map_err(|source| StoreError::Generic {
+                store: OBKV,
+                source: Box::new(source),
+            })?;
+
+        let mut common_prefixes = HashSet::new();
+        let mut objects = Vec::new();
+        for meta in metas {
+            let location = meta.location;
+            let subfix = &location[path.as_ref().len()..];
+            if let Some(pos) = subfix.find(DELIMITER) {
+                // common_prefix endswith '/'
+                let common_prefix = &subfix[0..pos + 1];
+                common_prefixes.insert(Path::from(common_prefix));
+            } else {
+                objects.push(ObjectMeta {
+                    location: Path::from(location),
+                    last_modified: Utc.timestamp_millis_opt(meta.last_modified).unwrap(),
+                    size: meta.size,
+                })
+            }
+        }
+
+        let common_prefixes = Vec::from_iter(common_prefixes.into_iter());
+        debug!(
+            "ObkvObjectStore list_with_delimiter operation, prefix:{path}, cost:{}",
+            instant.elapsed().as_millis()
+        );
+        Ok(ListResult {
+            common_prefixes,
+            objects,
+        })
+    }
+
+    async fn copy(&self, _from: &Path, _to: &Path) -> Result<()> {
+        // TODO:
+        Err(StoreError::NotImplemented)
+    }
+
+    async fn copy_if_not_exists(&self, _from: &Path, _to: &Path) -> Result<()> {
+        // TODO:
+        Err(StoreError::NotImplemented)
+    }
+}
+
+struct ObkvMultiPartUpload<T> {
+    /// The full path to the object.
+    location: Path,
+    /// The id of multi upload tasks, we use this id as object version.
+    upload_id: String,
+    /// The table name of obkv to save data.
+    table_name: String,
+    /// The client of object store.
+    client: Arc<T>,
+    /// The size of object.
+    size: AtomicU64,
+    /// The size in bytes of one part. Note: maybe the size of last part less
+    /// than part_size.
+    part_size: usize,
+    /// The mananger to process meta info.
+    meta_manager: Arc<MetaManager<T>>,
+}
+
+struct Prefix;
```

Comment:
Rename it as `PathKeyEncoder`, and provide the two methods:
- part_key
- prefix

Reply:
Ok, and there are 3 methods need to provide:
* part_key
* part_key_prefix
* unique_id